### ISSUE TITLE
`returnRecordToPojo: true` effectively disables reflection cache

### ISSUE THREAD

**Issue by schmeedy** [jOOQ/jOOQ#7731]
### Expected behavior and actual behavior:
Expected: `returnRecordToPojo` doesn't impact reflection caching 

Actual: if `returnRecordToPojo` is not set to `false`, new context might be derived for inserts / updates / deletes: https://github.com/jOOQ/jOOQ/blob/3146a29a19093ec82c214dfbcb85c9b09df14602/jOOQ/src/main/java/org/jooq/impl/DAOImpl.java#L411

This does not carry-over the context-level caches (including reflection caches).

**In case of many inserts:**

With `returnRecordToPojo == true`

![image](https://user-images.githubusercontent.com/645661/43829008-422f2202-9afe-11e8-9a92-607a944e9300.png)

With `returnRecordToPojo == false`

![image](https://user-images.githubusercontent.com/645661/43828698-88172414-9afd-11e8-8f7f-f1f362a37ced.png)

`Tools$9.call()` signifies a cache miss. In case of `returnRecordToPojo == true` there are only cache misses due to context-level reflection caches being discarded as described above

### Steps to reproduce the problem (include example code if possible):
Do some inserts with `returnRecordToPojo == true`

### Versions:
- jOOQ: 3.11

**Comment by lukaseder** [comment]
Thank you very much for your detailed report. Interesting, indeed there seems to be a low hanging fruit for optimisation. Will look into this soon

**Comment by lukaseder** [comment]
I'm currently looking into this. The "low hanging fruit" wasn't as low hanging as I originally thought. Maintaining caches across derived `Configuration` instances isn't trivial - there are many open questions and possible regressions. I have created a new issue for this: #8008

However, in your case, you're passing a large list of POJOs to `DAO.insert()`. The current implementation re-derives the `Configuration` on each iteration, in order to be able to pass the POJO reference to the `RecordListener` that collects the returned ID and maps it to the POJO. If we use e.g. an `IdentityHashMap<R, P>` that remembers the complete `Record` to `POJO` mapping, we don't have to create individual `RecordListener` instances, but can do with only one, referencing the entire map. That way, the lack of cache propagation is less painful, as the cache is re-created only once per loop, not per iteration.

Of course, this improvement doesn't help if the iteration is external to the `DAO.insert()` method, but I think it is already much better in your case.

**Comment by schmeedy** [comment]
@lukaseder thank you for looking into this! This fix should resolve our batch insert performance issue.

**Comment by lukaseder** [comment]
Excellent. Let me know if you find any other issues, or see other potential for performance improvements
